<h1 id="vypisy_chyb_ze_zanorenych_funkci">Výpisy chyb ze zanořených funkcí
<a href="#vypisy_chyb_ze_zanorenych_funkci" class="header-link">#</a>
</h1>
<p>Teď si ukážeme (nebo zopakujeme), jak Python vypíše chybu, která
nastane v zanořené funkci.</p>
<p>Pročti si následující ukázkový příklad.</p>
<div class="admonition note"><p>Tohle je absurdní ilustrační příklad, ne ukázka jak dobře programovat!</p>
</div><!-- XXX: automatic line numbers? -->

<div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">podel</span><span class="p">(</span><span class="n">delenec</span><span class="p">,</span> <span class="n">delitel</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Podělí čísla mezi sebou a vrátí výsledek.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">delenec</span> <span class="o">/</span> <span class="n">delitel</span>        <span class="c1"># řádek 3</span>


<span class="k">def</span> <span class="nf">podel_nulou</span><span class="p">(</span><span class="n">cislo</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Vydělí dané číslo nulou a vrátí výsledek.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">podel</span><span class="p">(</span><span class="n">cislo</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>          <span class="c1"># řádek 8</span>


<span class="k">def</span> <span class="nf">ukaz_priklad</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Spočítá ukázkový příklad a výsledek vypíše (nevrátí!)&quot;&quot;&quot;</span>
    <span class="n">vysledek</span> <span class="o">=</span> <span class="n">podel_nulou</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>       <span class="c1"># řádek 13</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">vysledek</span><span class="p">)</span>

<span class="n">ukaz_priklad</span><span class="p">()</span>                      <span class="c1"># řádek 16</span>
</pre></div><p>Co se stane, když tohle pustíš?</p>
<div class="highlight"><pre><span></span><span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;ukazka.py&quot;</span>, line <span class="m">16</span>, in <span class="n">&lt;module&gt;</span>
<span class="w">    </span><span class="n">ukaz_priklad</span><span class="p">()</span>
  File <span class="nb">&quot;ukazka.py&quot;</span>, line <span class="m">13</span>, in <span class="n">ukaz_priklad</span>
<span class="w">    </span><span class="n">vysledek</span> <span class="o">=</span> <span class="n">podel_nulou</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
  File <span class="nb">&quot;ukazka.py&quot;</span>, line <span class="m">8</span>, in <span class="n">podel_nulou</span>
<span class="w">    </span><span class="k">return</span> <span class="n">podel</span><span class="p">(</span><span class="n">cislo</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
  File <span class="nb">&quot;ukazka.py&quot;</span>, line <span class="m">3</span>, in <span class="n">podel</span>
<span class="w">    </span><span class="k">return</span> <span class="n">delenec</span> <span class="o">/</span> <span class="n">delitel</span>
<span class="gr">ZeroDivisionError</span>: <span class="n">division by zero</span>
</pre></div><p>Všimni si, že každá z funkcí, jejíž volání vedlo k chybě, je uvedena ve výpisu.
Skutečná chyba (tedy místo, které musíš opravit) je asi někde poblíž těchto
míst:</p>
<ul>
<li>Na řádku 3, ve funkci <code>podel</code>, vzniklo dělení nulou.
Měla by funkce <code>podel</code> zjistit, jestli dostane nulu, a nějak na to
zareagovat?</li>
<li>Na řádku 8, ve funkci <code>podel_nulou</code>, kde voláš <code>podel</code> s dělitelem 0.
Má to tak opravdu být?</li>
<li>Na řádku 13, ve funkci <code>ukaz_priklad</code>, kde voláš funkci <code>podel_nulou</code>.
Kdybys to nedělal/a, chyba by taky nevznikla!</li>
<li>Na řádku 16, ne ve funkci, když voláš funkci <code>ukaz_priklad</code>.
Možná by to chtělo použít jiný příklad?</li>
</ul>
<p>Python (a asi ani ty) nemůže vědět, co tím programem programátor myslel,
a kdy by tedy bylo nejlepší chybu opravit.
Ukáže tedy v programu všechna místa, která k chybě vedla.
Je na tobě, abys z nich vybral/a to nejvhodnější a zaměřil/a se na něj.</p>
<p>Tahle ukázka je samozřejmě jen teoretická, ale v reálných programech vypadá
hlášení chyb stejně.
Až takovou složitou hlášku uvidíš, nepanikař!
Python se snaží co nejvíc napovědět a usnadnit ti chybu najít a opravit.
Sice stroze a anglicky, ale snaží.
Vyjdi mu vstříc a nauč se tyhle hlášky číst.</p>